crates/executor/src/blockchain_tree/mod.rs

```diff
@@ -0,0 +1,811 @@
+use reth_primitives::{BlockHash, BlockNumber, ChainSpec, SealedBlock, SealedBlockWithSenders};
+use reth_provider::{
+    ExecutorFactory, HeaderProvider, ShareableDatabase, StateProvider, StateProviderFactory,
+    Transaction,
+};
+use std::{
+    collections::{BTreeMap, HashMap},
+    sync::Arc,
+};
+
+use self::block_indices::BlockIndices;
+
+#[cfg_attr(doc, aquamarine::aquamarine)]
+/// Tree of chains and its identifications.
+///
+/// Mermaid flowchart represent all blocks that can appear in blockchain.
+/// Green blocks belong to canonical chain and are saved inside database table, they are our main
+/// chain. Pending blocks and sidechains are found in memory inside [`BlockchainTree`].
+/// Both pending and sidechains have same mechanisms only difference is when they got committed to
+/// database. For pending it is just append operation but for sidechains they need to move current
+/// canonical blocks to BlockchainTree flush sidechain to the database to become canonical chain.
+/// ```mermaid
+/// flowchart BT
+/// subgraph canonical chain
+/// CanonState:::state
+/// block0canon:::canon -->block1canon:::canon -->block2canon:::canon -->block3canon:::canon --> block4canon:::canon --> block5canon:::canon
+/// end
+/// block5canon --> block6pending:::pending
+/// block5canon --> block67pending:::pending
+/// subgraph sidechain2
+/// S2State:::state
+/// block3canon --> block4s2:::sidechain --> block5s2:::sidechain
+/// end
+/// subgraph sidechain1
+/// S1State:::state
+/// block2canon --> block3s1:::sidechain --> block4s1:::sidechain --> block5s1:::sidechain --> block6s1:::sidechain
+/// end
+/// classDef state fill:#1882C4
+/// classDef canon fill:#8AC926
+/// classDef pending fill:#FFCA3A
+/// classDef sidechain fill:#FF595E
+/// ```
+///
+///
+/// main functions:
+/// * insert_block: insert block inside tree. Execute it and save it to database.
+/// * finalize_block: Flush chain that joins to finalized block.
+/// * make_canonical: Check if we have the hash of block that we want to finalize and commit it to
+///   db. If we dont have the block pipeline syncing should start to fetch the blocks from p2p.
+/// *
+/// Do reorg if needed
+
+pub struct BlockchainTree<DB: Database, C: Consensus, EF: ExecutorFactory> {
+    /// chains and present data
+    pub chains: HashMap<ChainId, Chain>,
+    /// Static chain id generator
+    pub chain_id_generator: u64,
+    /// Indices to block and their connection.
+    pub block_indices: BlockIndices,
+    /// Number of block after finalized block that we are storing. It should be more then
+    /// finalization window
+    pub num_of_side_chain_max_size: u64,
+    /// Finalization windows. Number of blocks that can be reorged
+    pub finalization_window: u64,
+    /// Externals
+    pub externals: Externals<DB, C, EF>,
+}
+
+/// Container for external abstractions.
+pub struct Externals<DB: Database, C: Consensus, EF: ExecutorFactory> {
```

Comment:
I think this should be private as it is an implementation detail here and the name doesn't mean much out of context